### DISCUSSION TITLE
Is it possible to add libs dynamically?

### DISCUSSION THREAD

**Discussion by addlistener** [pyinstaller/pyinstaller#7232] [Help] [answered]
**Option1**
Bundle an interpreter, download matplotlib (or whatever other library) and run the source code. But what if there's C++ code in the lib? is it portable? 

**Option2**
Bundle pip as well? I don't know if this is possible, since we run pyinstaller generated distributables on the client machine without a python environment.

**Background**
I'm trying to evaluate if it's possible to make a portable version of python to support our software. I figured that pyinstaller is making the cpython build process smooth and would love to discuss on it's limits.

**Comment by bwoodsend** [comment, marked as answer]
Ughh, this question again. The short answer is no, a PyInstaller build is immutable. See [any conversation about plugins](https://github.com/orgs/pyinstaller/discussions?discussions_q=plugins) for the long answer.

**Reply by addlistener** [reply]
Thanks. I did the search but didn't spot this question

**Comment by addlistener** [comment]
I actually found sth interesting.

If i use venv and bundle all the `venv/lib/site-packages` together in to `dist/{scriptname}` folder, my scripts runs! I know it might not be universal, but that means maybe we can add source code (and the dependency of the module) to the `dist/{scriptname}` folder after the artifact has been distributed?

**Reply by rokm** [reply]
Well, that's how `sys.path` works in python... (and top-level application directory is added to `sys.path`). So you can either copy extra packages into top-level application directory, or to some arbitrary directory and add path to that directory to `sys.path`. But you'll need to ensure that you copied all dependencies. So it is not impossible, if you know what you are doing. But we offer no support for such endeavors because they are situation-specific and out of scope of PyInstaller, which is why the short answer given earlier was "no".

**Reply by addlistener** [reply]
Thanks for explaining. I know tons of questions is overwhelming. But I'd love to say you guys have the best support I've ever met in open source projects.

I guess for me it's like utilizing the bootloader to offer cross platform building and running the source code directly.

**Reply by rokm** [reply]
In that case, it's probably best idea to ensure that whole python stdlib is collected (see #6824). And probably also that as little of anything else as possible is collected (or that whatever is collected, is collected in full); lest you run in conflicts between partially-collected packages in PYZ and your full extra copies (see https://github.com/pyinstaller/pyinstaller/discussions/7214#discussioncomment-4045999); although I *think* this would be a potential issue only when external location is used for extra packages, not with top-level application directory.
